Add SVG to the .zip (redo #897)

tomalec opened the pull request:
### Changes proposed in this Pull Request:

Use `file-loader` to expose image assets from `js/src/` to `js/build`.
Redo https://github.com/woocommerce/google-listings-and-ads/pull/897/

Fixes https://github.com/woocommerce/google-listings-and-ads/issues/498
and adds `.svg`s to the `.zip`  package https://github.com/woocommerce/google-listings-and-ads/pull/897#issuecomment-888246189

### Screenshots:

<!--- Optional --->


### Detailed test instructions:

1. `npm run build`
2. Redo tests from https://github.com/woocommerce/google-listings-and-ads/pull/897#issue-694550944

### Additional notes:

1. This PR tries to implement the idea to use Webpack's `file-loader` suggested at https://github.com/woocommerce/google-listings-and-ads/pull/897#issuecomment-888246189

	It almost works, but, for some reason `file-loader` cannot pickup `__webpack_public_path__` and the paths gets rendered as `images/{path}/{hash}.{name}.{ext}` whereas we need `{path_to_build_folder}/images/{path}/{hash}.{name}.{ext}`
	
	Is that achievable via Webpack, to do we need to add it as before with `glaData.assetsURL`?
	
	@ecgan @eason9487 Do you have any advice?

2. We could tweak the `file-loader`'s setup a little and make it will serve the files directly from `js/src/` folder for the development environment. So, we will not need to rebuild the package when files change. I'm not sure how often we would use that. If it's worth the bloat in the config, and the fact we will be testing not the production result.

### Changelog entry

tomalec committed:
Use `file-loader` to expose image assets from `js/src/` to `js/build`.

Redo https://github.com/woocommerce/google-listings-and-ads/pull/897/

Fixes https://github.com/woocommerce/google-listings-and-ads/issues/498
and adds `.scg`s to the `.zip`  package https://github.com/woocommerce/google-listings-and-ads/pull/897#issuecomment-888246189

ecgan commented:
> @ecgan @eason9487 Do you have any advice?

Unfortunately I don't. I'm not really an expert in webpack things. If you like, I can help look into this probably next week, after I'm done with porter rotation this week. Hope @eason9487 has some ideas on this. 😄

tomalec committed:
Dynamically set `__webpack_public_path__`,

to make `file-loader` resolve full URLs accurately.
We could make it use a server-side set variable like WooCommerce-admin does https://github.com/woocommerce/woocommerce-admin/blob/391c99df5accc2e1dda53e70abc686c85cd4ca26/client/index.js#L21.
But this way we are more flexible, as we don't tie a hidden link between the Webpack config and our PHP code.

Preferably, I'd make Webpack set it automatically https://webpack.js.org/guides/public-path/#automatic-publicpath,
but that is available only with Webpack 5.

tomalec commented:
I found out we can dynamically set `__webpack_public_path__`. by just executing: `__webpack_public_path__ = document.currentScript.src` in the main `index.js`


We could make it use a server-side set variable like WooCommerce-admin does https://github.com/woocommerce/woocommerce-admin/blob/391c99df5accc2e1dda53e70abc686c85cd4ca26/client/index.js#L21.
But this way we are more flexible, as we don't tie a hidden link between the Webpack config and our PHP code.

Preferably, I'd make Webpack set it automatically https://webpack.js.org/guides/public-path/#automatic-publicpath,
but that is available only with Webpack 5.

Other downsides I see here are that `document.currentScript` may not work for
- native apps,
- SSR
if we ever target those.

eason9487 commented on webpack.config.js:
```diff
@@ -31,9 +31,32 @@ const requestToHandle = ( request ) => {
+				use: {
+					loader: 'file-loader',
+					options: {
+						name: 'images/[path]/[contenthash].[name].[ext]',
```
💅 
Consider it takes the `[contenthash]`, maybe we could use `'images/[contenthash].[name].[ext]'` without `[path]` here. Thus, it would have a bit benefits:

- If we (accidentally or intentionally) have duplicate SVG files under `js/src`, it'd only emit one SVG to `js/build/images` due to they have the same content hash. Thus, the archive zip file size would not increase the duplicate parts.
- It doesn't affect much, but the final URL could be shorter, and no need to have the full subpath under `js/build/images`.

tomalec replied:
> If we (accidentally or intentionally) have duplicate SVG files under `js/src`, it'd only emit one SVG to `js/build/images` due to they have the same content hash. Thus, the archive zip file size would not increase the duplicate parts.

I was thinking about it as saving package save and de-duplicating is a valuable gain. But to achieve that we would have to either the same filename or reduce it to `'images/[contenthash].[ext]'`. Which then impairs debugging, development, and extremely niche user experience if someone happens to download those files. The last one could be ignored for our plugin, but I think a debugging case could be valuable.

Imagine a case, when we have a merchant reporting some issue to a Happiness Engineer, and HE discovered a `404 /images/afu76poo0a0602bca20d01a0ff048516339987a8ad1` in logs. To resolve the file to even know whether the 404 is something related to the issue or not, they need either guess, run some tooling, run the build for the exact revision of the plugin, etc.

If the log looks like `404 /images/js/src/components/account-card/afu76poo0a0.gogole-g-logo.svg`, they could more easily judge
- "Ok, this is not related to our issue of Automatewoo not triggering…"
- "Ah, so the image file is not there, could you check it in the file system", or "Maybe you happen to block URLs with bad words, and `poo` got caught"

> It doesn't affect much, but the final URL could be shorter, and no need to have the full subpath under `js/build/images`.


Personally, I considered full path a DevX as it allows a developer to match 1:1 the files used on production build to the file in their source.

To simplify/shorten the file path we could try to remove `/js/src` from there. So to make `/js/src/components/account-card/index.js` & `/js/src/components/account-card/gogole-g-logo.svg`, become `/js/build/index.js` & `/js/build/images/components/account-card/gogole-g-logo.svg` (instead of `/js/build/images/js/src/components/account-card/gogole-g-logo.svg`) .
But I'm not sure how easy would it be.

tomalec replied:
To me, it looks like a trade-off between
- de-duplication, vs.
- debugging experience

WDYT?

eason9487 replied:
Thanks for sharing your thoughts. I believe both of these scenarios will happen very rarely. But once if any of them occur, definitely the debugging experience brings more benefit.

eason9487 reviewed (approved):
LGTM and left a comment that might be polished a little bit.
